### ISSUE TITLE
lib_exec and other dynamic workarounds broken

### ISSUE THREAD

**Issue by gisikw** [KSP-KOS/KOS#1273]
Latest version appears to be caching even .ksm compiled files, which removes the ability to run any files/functions/code dynamically.

## Using [lib_exec](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_exec.ks)
```
RUN lib_exec.

UNSET x.
EXECUTE("SET x TO 5.").
PRINT x.
```
Output: *Undefined Variable Name \`x\`*

## Manual Recompile
```
LOG " " TO otherFile.ksm.
DELETE otherFile.ksm.
LOG " " TO otherFile.ks.
DELETE otherFile.ks.

LOG "PRINT 1." TO otherFile.ks.
COMPILE otherFile.
DELETE otherFile.ks.
RUN otherFile.

LOG " " TO otherFile.ksm.
DELETE otherFile.ksm.
LOG " " TO otherFile.ks.
DELETE otherFile.ks.

LOG "PRINT 2." TO otherFile.ks.
COMPILE otherFile.
DELETE otherFile.ks.
RUN otherFile.
```
Output:
```
1
1
```

Similar errors with lib_exec's `EVALUATE_FUNCTION`, and unfortunately, I can't seem to find a workaround.

**Comment by hvacengi** [comment]
The best work around I can think of would be to use `random()` to seed a file name (like name the file `"compile" + round(random()*10000) + ".ksm"`). The issue is that the name is being cached.  While this was true in past versions as well, there was a bug fix that must have affected this library.

Honestly, I've always thought this was a fairly hacky solution.  If this "feature" is getting used that frequently, it would be far better to provide a built in string evaluation.  It can't realistically happen until we look at escaped strings, so that you can use `"` within the string.  More importantly however, @tomekpiotrowski is working on serialization and deserialization to allow for data contracts.  I'm also working on a way to include a script block in the craft's editor description.

As near as I can tell, this is most useful for trying to store values in nametags.  What are other common use cases?

**Comment by gisikw** [comment]
The problem is that even generating a random file name doesn't allow it to be executed. I can generate "foo531531.ksm, but can't `RUN "foo531531.ksm"`, so it just sorta kicks the can down the road. This was manageable when lib_exec still worked, as you could `LOG "RUN " + randomFilename + ".ksm" TO fileRunner.ks.`, and lib_exec fileRunner, but that was more of a minification optimization than anything else.

As far as use cases, I know that a few folks on the subreddit were also relying on recompiling (using KSM compilation to cache-bust), but here are my own use cases, off the top of my head:

* Dynamic function calls `LOG "reactTo" + SHIP:STATE + "().".` (also provides a workaround to allow higher-order functions, as the function name can be passed in as an argument, in lieu of something like `FOO(&BAR).`)
* Dynamic file calls `LOG "RUN reactTo" + SHIP:STATE + ".ks.".`
* Caching of reboot-persistent data in any sort of form, via string serialized lists, lexicons, vars, etc.

For the *most* part, my own use cases could be replaced by:
* Syntax for functions as arguments/return values (`FUNCTION FOO(&BAR) { RETURN &BAZ. }`)
* Non-cached File -> String, now that the String manipulation methods exist (`SET myData TO READ("pidValues.csv"):split(",").`)

lib_exec served as a workaround to provide this sort of functionality, albeit with some overhead. But I do know that others were relying on force-recompilation strategies on the subreddit, and can't speak to their exact use cases.

**Comment by Dunbaratu** [comment]
I'm calling it `enhancement` because as annoying as the current behavior is, it is how the system is designed.  Fixing it requires some fancy changes that I don't mind doing eventually, for other reasons.

This is, essentially, a side effect of the issue mentioned in #1257.  When it gets addressed, that should end up addressing this too.

**Comment by gisikw** [comment]
Yep, agreed. Once we have dynamic `RUN`, there should never be any need to try and force a way around caching. Though lib_exec will probably need a rewrite at some point once that's in place.
